feat: BROS-44: Add MultiChannel tag support for TimeSeries

Gondragos opened the pull request:
Add the ability to group channels into one plot.

### Before
<img width="855" alt="image" src="https://github.com/user-attachments/assets/6454d78b-1ccc-4f26-a223-f3dfc5763556" />

### After
![Screen Recording 2025-06-02 at 07 41 47](https://github.com/user-attachments/assets/23238ff0-83de-435a-998b-8daf1e0276cb)


---

## Summary of the Pull Request

This PR introduces key updates designed to enhance the functionality of the TimeSeries component through new features.

## Feature Flags:

Added FF_MULTICHANNEL_TS feature flag to enable multi-channel visualization for TimeSeries.

## New Features:
- Introduction of TimeSeriesVisualizer Component: A React-based D3-powered visualization component to support single-channel and multi-channel TimeSeries data rendering. Based on Channel renderer with the addition of the ability to handle more than one channel at the same plot.
- MultiChannel Support: Added MultiChannel tag for grouped rendering of multiple channels in TimeSeries with interactions for individual channel highlighting, hiding, and customization (ChannelLegend added for UI).
- Add legend to MultiChanel, that allows highlighting and hiding specific channels.
- Dynamic adjusting default colors for channels based on their index in `MultiChannel` or `Timeseries`
## Feature Enhancements:
- Modified TimeSeries tag and Channel to support grouped multi-channel visualization.
- Added new computed properties (channelsMap, channels) for efficient channel management.

## Rollout Strategy: 
The critical changes are gated under the FF_MULTICHANNEL_TS feature flag.

<!--

This description MUST be filled out for a PR to receive a review. Its primary purposes are:

 - to enable your reviewer to review your code easily, and
 - to convince your reviewer that your code works as intended.

Some pointers to think about when filling out your PR description:
 - Reason for change: Description of problem and solution
 - Screenshots: All visible changes should include screenshots.
 - Rollout strategy: How will this code be rolled out? Feature flags / env var / other
 - Testing: Description of how this is being verified
 - Risks: Are there any known risks associated with this change, eg to security or performance?
 - Reviewer notes: Any info to help reviewers approve the PR
 - General notes: Any info to help onlookers understand the code, or callouts to significant portions.

You may use AI tools such as Copilot Actions to assist with writing your PR description (see https://docs.github.com/en/copilot/using-github-copilot/using-github-copilot-for-pull-requests/creating-a-pull-request-summary-with-github-copilot); however, an AI summary isn't enough by itself. You'll need to provide your reviewer with strong evidence that your code works as intended, which requires actually running the code and showing that it works.

-->

Gondragos committed:
feat: BROS-44: Add MultiChannel tag support for TimeSeries

- Introduced a feature flag `FF_MULTICHANNEL_TS` to enable multi-channel functionality in the TimeSeries component.
- Added new tag `MultiChannel` for grouping channels on one plot.
- Added `ChannelLegend` and `TimeSeriesVisualizer` components for improved visualization and interaction.
- Updated the `Channel` model and several associated functionalities to support multi-channel rendering.

Gondragos committed:
test: Add small smoke test

Gondragos committed:
Linting

Gondragos committed:
feat: Implement dynamic color assignment for channels

- Introduced a palette of color tokens for consistent styling.
- Added functionality to assign colors dynamically to stroke and marker properties of channels.
- Updated `MultiChannel` and `TimeSeries` components to utilize the new dynamic color logic.

niklub committed:
Merge branch 'develop' into fb-BROS-44/multichannel

hlomzik commented on web/libs/editor/src/tags/object/TimeSeries/MultiChannel.jsx:
```diff
@@ -0,0 +1,95 @@
+    parentTypes: Types.tagsTypes(["TimeSeries"]),
+    height: types.optional(types.string, "200"),
+
+    showaxis: types.optional(types.boolean, true),
```
shouldn't it be consistent?
```suggestion
    showxaxis: types.optional(types.boolean, true),
```

Gondragos replied:
nope. it's for both.

Gondragos commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/15478123181)
> Successfully merged:  rename web/libs/{datamanager/src => core/src/lib}/utils/urlJSON.ts (100%)

Gondragos committed:
Merge branch 'develop' into 'fb-BROS-44/multichannel'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/15478123181

Gondragos committed:
fix: Fix TS visualizer for single channel

Gondragos committed:
fix: Correct TimeSeries signal generation logic

Allow to collect all columns

Gondragos committed:
fix: Prevent ChannelLegend from triggering seek action

Gondragos committed:
fix: Ensure default valueType is "url" in TimeSeries

Added fallback logic to set `value_type` as "url" when not specified for TimeSeries. This ensures proper handling of cases where the `valueType` is undefined, improving consistency in processing.

Gondragos committed:
fix: Respect `fixedscale` property in MultiChannel

Updated `MultiChannel` and `TimeSeriesVisualizer` components to correctly respect the `fixedscale` property when defined at various levels. Added logic to handle inheritance of the `fixedscale` value to ensure proper scaling behavior within nested structures.

Gondragos committed:
fix: Replace equality check with `is None`

Corrected the equality check for `value_type` from `== None` to `is None`. This change ensures adherence to Python best practices and improves code reliability.

Gondragos committed:
docs: Add documentation for MultiChannel tag

First attempt into docs

Gondragos committed:
docs: Add params

Gondragos committed:
Merge branch 'develop' of ssh://github.com/heartexlabs/label-studio into develop

# Conflicts:
#	web/libs/core/src/lib/utils/feature-flags/flags.ts

Gondragos committed:
fix: Add playhead rendering to TimeSeriesVisualizer component

Introduced the `playhead` feature (previously defined in Channel) for visualizing cursor position during playback in `TimeSeriesVisualizer`. Added methods to render, update, and sync the playhead position, ensuring proper alignment with scale changes and cursor time updates. Updated `Channel` and `MultiChannel` components to pass the `cursorTime` prop for synchronization.

Gondragos committed:
docs: Update documentation for MultiChannel tag

Added `@name` annotation for the `MultiChannel` component and refined parameter descriptions for visibility in docs builder

robot-ci-heartex committed:
ci: Build tag docs

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/15492103234

Gondragos committed:
docs: Consolidate MultiChannel documentation into TimeSeries

Moved MultiChannel tag documentation from a standalone file to the TimeSeries tag documentation for better organization. Updated related examples and references to ensure consistency across the documentation.

Gondragos committed:
Merge remote-tracking branch 'origin/fb-BROS-44/multichannel' into fb-BROS-44/multichannel

Gondragos committed:
docs: Fix docs

robot-ci-heartex committed:
ci: Build tag docs

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/15492512614

Gondragos committed:
fix: Adjust default valueType behavior in secure mode

Updated the logic to determine default `valueType` for certain tags (`Paragraphs`, `HyperText`, `Text`) in secure mode, assigning it as "url" when not explicitly defined. Removed redundant handling of `valueType` for `TimeSeries` to streamline the validation process and avoid duplication.

Gondragos committed:
Merge remote-tracking branch 'origin/fb-BROS-44/multichannel' into fb-BROS-44/multichannel

Gondragos committed:
docs: Expand TimeSeries documentation for MultiChannel

Added detailed documentation for the MultiChannel tag, including its features, usage, and components like ChannelLegend, TimeSeriesVisualizer, and the color palette system. Updated examples to reflect new functionality and replaced references to legacy `ChannelD3` with `TimeSeriesVisualizer` for consistency.

makseq committed:
Fix margin that caused shift in cursor position while manual navigation

makseq committed:
Fix for video and timeseries navigation while playback

makseq committed:
Linter

Gondragos commented on web/libs/editor/src/components/TimeSeries/TimeSeriesVisualizer.jsx:
```diff
@@ -404,7 +404,8 @@ class TimeSeriesVisualizerD3 extends React.Component {
     const { width } = this.state;
-    const { margin } = item;
+    // Use parent TimeSeries margin for consistent coordinate system
+    const margin = item.parent?.margin || item.margin;
```
Using `item.margin` was intended as `item.parent?.margin` does not contain margin for the second Y axis in multi-channel variant.

Gondragos commented on web/libs/editor/src/tags/object/TimeSeries/README.md:
```diff
@@ -132,7 +132,24 @@ Identical logic but uses scaled brush coordinate.
+* **Result**: Both video and TimeSeries cursor jump to clicked position and continue playing synchronously.
+
+### 4.6 Shared click handling logic
+* `handleTimeSeriesMainAreaClick()` in `helpers.js` provides unified click handling for both:
```
I don't think we need that as `handleMainAreaClick` worked in both places already.

Gondragos commented on web/libs/editor/src/components/TimeSeries/TimeSeriesVisualizer.jsx:
```diff
@@ -943,7 +947,7 @@ class TimeSeriesVisualizerD3 extends React.Component {
     );
 
-    return <div className="htx-timeseries-channel" ref={this.ref} />;
+    return <div className="htx-timeseries-channel" ref={this.ref} onClick={this.handleMainAreaClick} />;
```
Using such handler here most probably will cause calling `handleTimeSeriesMainAreaClick` twice.

Gondragos committed:
Revert unnecessary changes

Gondragos committed:
Revert "Fix margin that caused shift in cursor position while manual navigation"

This reverts commit 86f858f5b54cbbee9cc7a64757d95ffaf557d3a5.

Gondragos committed:
feat: Add plot click handling to TimeSeries

Use internal d3 values directly.

Gondragos commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/15525884630)
> Successfully merged:  8 files changed, 95 insertions(+), 34 deletions(-)

Gondragos committed:
Merge branch 'develop' into 'fb-BROS-44/multichannel'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/15525884630

hlomzik commented on web/libs/editor/src/tags/object/TimeSeries.jsx:
```diff
@@ -164,6 +178,24 @@ const Model = types
       return self.timeformat && self.timecolumn ? d3.utcParse(self.timeformat) : Number;
     },
 
+    get channelsMap() {
```
I believe you could use something like `traverseTree(self, item => item.type === "channel" && res[item.columnName] = item)`
But this would work too of course

Gondragos replied:
🤯  I completely forgot about that

bmartel commented on web/libs/editor/src/tags/object/TimeSeries/ChannelLegend.module.scss:
```diff
@@ -0,0 +1,31 @@
+.channelLegend {
+  --marker-size: 10px;
+  --marker-color:  var(--color-primary-surface, #4C5FA9);
+  --marker-hidden-color: var(--color-neutral-surface, #F9F8F6);
```
It's better to not default semantic tokens as we won't know up front if the correct dark mode compatible colours would be applying.

```suggestion
  --marker-color:  var(--color-primary-surface);
  --marker-hidden-color: var(--color-neutral-surface);
```

Gondragos replied:
How can we have a dark mode without setting its own values?

bmartel replied:
The defaults here are not required and also can cause issues if there were to be styles that were not getting applied correctly, thus it is safer to not set defaults to CSS vars, since the defaults are not dark mode compatible themselves and would mask problems which we would otherwise visually see very simply in a review of the running feature. We have already run into that a bunch in the work done to make the application dark mode compatible and theme ready, which makes it hard to spot what in particular is wrong.

makseq committed:
Add delta to generate_time_series_json and set second as default delta

bmartel commented on web/libs/editor/src/tags/object/TimeSeries/palette.js:
These colours won't be dark mode compatible as they are from the primitive colour collections.

We have a dark mode compatible collection `accent` of the colours here:

https://labelstud.io/storybook/?path=/story/design-tokens--tokens-catalog

We should likely aim for `base` of each color type, with sand using `bold`

bmartel commented on web/libs/editor/src/tags/object/TimeSeries/palette.js:
```diff
@@ -0,0 +1,56 @@
+// Color tokens from Figma
+const colorTokens = {
+  grape: {
+    400: "var(--color-grape-400)",
+    hex: "#6D87F1",
```
```suggestion
    base: "var(--color-accent-grape-base)",
```

bmartel commented on web/libs/editor/src/tags/object/TimeSeries/palette.js:
```diff
@@ -0,0 +1,56 @@
+  },
+  blueberry: {
+    400: "var(--color-blueberry-400)",
+    hex: "#5B8AF4",
```
```suggestion
    base: "var(--color-accent-blueberry-base)",
```

bmartel commented on web/libs/editor/src/tags/object/TimeSeries/palette.js:
```diff
@@ -0,0 +1,56 @@
+  },
+  kale: {
+    400: "var(--color-kale-400)",
+    hex: "#36B37E",
```
```suggestion
    base: "var(--color-accent-kale-base)",
```

bmartel commented on web/libs/editor/src/tags/object/TimeSeries/palette.js:
```diff
@@ -0,0 +1,56 @@
+  },
+  kiwi: {
+    400: "var(--color-kiwi-400)",
+    hex: "#79F2C0",
```
```suggestion
    base: "var(--color-accent-kiwi-base)",
```

bmartel commented on web/libs/editor/src/tags/object/TimeSeries/palette.js:
```diff
@@ -0,0 +1,56 @@
+  },
+  mango: {
+    400: "var(--color-mango-400)",
+    hex: "#FFAB00",
```
```suggestion
    base: "var(--color-accent-mango-base)",
```

bmartel commented on web/libs/editor/src/tags/object/TimeSeries/palette.js:
```diff
@@ -0,0 +1,56 @@
+    hex: "#FFAB00",
+  },
+  canteloupe: {
+    400: "var(--color-canteloupe-400)",
+    hex: "#FF8B00",
```
```suggestion
    base: "var(--color-accent-canteloupe-base)",
```

bmartel commented on web/libs/editor/src/tags/object/TimeSeries/palette.js:
```diff
@@ -0,0 +1,56 @@
+    hex: "#FF8B00",
+  },
+  persimmon: {
+    400: "var(--color-persimmon-400)",
+    hex: "#FF5630",
```
```suggestion
    base: "var(--color-accent-persimmon-base)",
```

bmartel commented on web/libs/editor/src/tags/object/TimeSeries/palette.js:
```diff
@@ -0,0 +1,56 @@
+  },
+  plum: {
+    400: "var(--color-plum-400)",
+    hex: "#8777D9",
```
```suggestion
    base: "var(--color-accent-plum-base)",
```

bmartel commented on web/libs/editor/src/tags/object/TimeSeries/palette.js:
```diff
@@ -0,0 +1,56 @@
+  },
+  fig: {
+    400: "var(--color-fig-400)",
+    hex: "#6554C0",
```
```suggestion
    base: "var(--color-accent-fig-base)",
```

bmartel commented on web/libs/editor/src/tags/object/TimeSeries/palette.js:
```diff
@@ -0,0 +1,56 @@
+  },
+  sand: {
+    700: "var(--color-sand-700)",
+    hex: "#505F79",
```
```suggestion
    bold: "var(--color-accent-sand-bold)",
```

bmartel commented on web/libs/editor/src/tags/object/TimeSeries/palette.js:
```diff
@@ -0,0 +1,56 @@
+ */
+export const getChannelColor = (index) => {
+  const colorName = colorOrder[index % colorOrder.length];
+  return colorTokens[colorName][colorName === "sand" ? "700" : "400"];
```
```suggestion
  return colorTokens[colorName][colorName === "sand" ? "bold" : "base"];
```

bmartel reviewed (changes requested):
Just the colours need to be updated to be dark mode compatible which I left suggestions for each, rest looks awesome!

Commit:
Apply suggestions from code review

Fix colors

Co-authored-by: bmartel <brandonmartel@gmail.com>

Gondragos commented:
/fmt
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/15535094324)

Gondragos committed:
Apply pre-commit linters

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/15535094324

makseq committed:
Add new generation for timeseries

makseq committed:
Merge branch 'fb-BROS-44/multichannel' of github.com:heartexlabs/label-studio into fb-BROS-44/multichannel

Gondragos committed:
fix: Correct import path for FF_MULTICHANNEL_TS
